### PR TITLE
Make account endorsements idempotent (fix #19045)

### LINKED ISSUE #19045: API: api/v1/accounts/:id/pin should be idempotent
### Steps to reproduce the problem

1. Try to endorse an account you are already endorsing

### Expected behaviour

Relationship entity is returned with 200 OK

### Actual behaviour

In some versions you get 422 with "Duplicate record", in other versions you get a 500 error

### Specifications

mastodon.social (currently 422)

### PR DESCRIPTION
**Pull request by trwnh**
Fix #19045

### CODE REVIEW

File: app/controllers/api/v1/accounts/pins_controller.rb [hunk 1]
```diff
@@ -8,7 +8,10 @@ class Api::V1::Accounts::PinsController < Api::BaseController
   before_action :set_account
 
   def create
-    AccountPin.create!(account: current_account, target_account: @account)
+    pin = AccountPin.find_by(account: current_account, target_account: @account)
+    if pin.nil?
```
Comment (ykzts) [inline app/controllers/api/v1/accounts/pins_controller.rb:12]:
```suggestion
    if AccountPin.exists?(account: current_account, target_account: @account)
```

Comment (trwnh) [reply app/controllers/api/v1/accounts/pins_controller.rb]: Thank you!

File: app/controllers/api/v1/accounts/pins_controller.rb [hunk 2]
```diff
@@ -8,8 +8,7 @@ class Api::V1::Accounts::PinsController < Api::BaseController
   before_action :set_account
 
   def create
-    pin = AccountPin.find_by(account: current_account, target_account: @account)
-    if pin.nil?
+    if AccountPin.exists?(account: current_account, target_account: @account)
```
Comment (christophsturm) [inline app/controllers/api/v1/accounts/pins_controller.rb:11]: `unless` instead of `if`

**Review by christophsturm — CHANGES_REQUESTED** [review]

**Review by ykzts — APPROVED** [review]

File: app/controllers/api/v1/accounts/pins_controller.rb [hunk 3]
```diff
@@ -8,7 +8,9 @@ class Api::V1::Accounts::PinsController < Api::BaseController
   before_action :set_account
 
   def create
-    AccountPin.create!(account: current_account, target_account: @account)
+    unless AccountPin.exists?(account: current_account, target_account: @account)
+      AccountPin.create!(account: current_account, target_account: @account)
+    end
```
Comment (ClearlyClaire) [inline app/controllers/api/v1/accounts/pins_controller.rb:13]: `AccountPin.find_or_create_by!`? It's functionally equivalent (that is, still not doing the change atomically) but more idiomatic

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**